WordPress/Classes/Models/Blog+Lookup.swift

```diff
@@ -58,4 +69,79 @@ public extension Blog {
 
         return service.findBlog(withXmlrpc: xmlrpc, andUsername: username)
     }
+
+    @objc(countInContext:)
+    static func count(in context: NSManagedObjectContext) -> Int {
+        BlogQuery().count(in: context)
+    }
+
+    @objc(wpComBlogCountInContext:)
+    static func wpComBlogCount(in context: NSManagedObjectContext) -> Int {
+        BlogQuery().hostedByWPCom(true).count(in: context)
+    }
+
+    @objc(selfHostedInContext:)
+    static func selfHosted(in context: NSManagedObjectContext) -> [Blog] {
+        (try? BlogQuery().hostedByWPCom(false).blogs(in: context)) ?? []
+    }
+}
+
+struct BlogQuery {
+    private var predicates = [NSPredicate]()
+
+    func blogID(_ id: Int) -> Self {
+        blogID(Int64(id))
+    }
+
+    func blogID(_ id: NSNumber) -> Self {
+        blogID(id.int64Value)
+    }
+
+    func blogID(_ id: Int64) -> Self {
+        and(NSPredicate(format: "blogID = %ld", id))
+    }
+
+    func username(_ username: String) -> Self {
+        and(NSPredicate(format: "account.username = %@", username))
+    }
+
+    func hostname(_ hostname: String) -> Self {
+        and(NSPredicate(format: "url CONTAINS %@", hostname))
+    }
+
+    func visible(_ flag: Bool) -> Self {
+        and(NSPredicate(format: "visible = %@", NSNumber(booleanLiteral: flag)))
+    }
+
+    func hostedByWPCom(_ flag: Bool) -> Self {
+        and(NSPredicate(format: flag ? "account != NULL" : "account == NULL"))
+    }
+
+    func count(in context: NSManagedObjectContext) -> Int {
+        (try? context.count(for: buildFetchRequest())) ?? 0
+    }
+
+    func blog(in context: NSManagedObjectContext) throws -> Blog? {
+        let request = buildFetchRequest()
+        request.fetchLimit = 1
+        return (try context.fetch(request).first)
+    }
+
+    func blogs(in context: NSManagedObjectContext) throws -> [Blog] {
+        try context.fetch(buildFetchRequest())
+    }
+
+    private func buildFetchRequest() -> NSFetchRequest<Blog> {
+        let request = NSFetchRequest<Blog>(entityName: Blog.entityName())
+        request.includesSubentities = false
+        request.sortDescriptors = [NSSortDescriptor(key: "settings.name", ascending: true)]
+        request.predicate = NSCompoundPredicate(andPredicateWithSubpredicates: predicates)
+        return request
```

Comment:
[The original query method](https://github.com/wordpress-mobile/WordPress-iOS/blob/9f436490c917099aba8dfb0ebf695305475523e5/WordPress/Classes/Services/BlogService.m#L832-L869):
```objc
- (NSArray *)blogsWithPredicate:(NSPredicate *)predicate
{
    NSFetchRequest *request = [self fetchRequestWithPredicate:predicate];
    NSSortDescriptor *sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"settings.name"
                                                                     ascending:YES];
    request.sortDescriptors = @[ sortDescriptor ];
    // ......
}

- (NSFetchRequest *)fetchRequestWithPredicate:(NSPredicate *)predicate
{
    NSFetchRequest *request = [NSFetchRequest fetchRequestWithEntityName:NSStringFromClass([Blog class])];
    request.includesSubentities = NO;
    request.predicate = predicate;
    return request;
}
```